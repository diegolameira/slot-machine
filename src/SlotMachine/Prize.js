import React from 'react';

export const calcPrize = (items = []) => {
  let equals = 0;
  let consecutives = false;
  items.reduce((prev, current, index, array) => {
    const compare = [
      ...array.slice(0, index),
      ...array.slice(index + 1, array.length),
    ];
    if (prev === current) consecutives = true;
    if (~compare.indexOf(current)) equals++;
    return current;
  }, '');
  return equals > 0 && equals === items.length
    ? 100
    : equals && consecutives
    ? 20
    : equals
    ? 10
    : 0;
};

export default ({ history = [] }) => (
  <>
    <h1>Prize: {calcPrize(history.slice(-1)[0])}</h1>
    <h2>
      Win Amount:
      {history.reduce((prev, current) => {
        return calcPrize(current) + prev;
      }, 0)}
    </h2>
  </>
);
