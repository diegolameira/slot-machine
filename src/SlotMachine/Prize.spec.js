import React from 'react';
import TestRenderer from 'react-test-renderer';

import Prize, { calcPrize } from './Prize';

describe('Prize', () => {
  let history, component;

  beforeAll(() => {
    history = [];
    component = TestRenderer.create(<Prize history={history} />);
  });

  beforeEach(() => {
    component.update(<Prize history={history} />);
  });

  afterEach(() => {
    expect(component.toJSON()).toMatchSnapshot();
  });

  history = [[1, 0, 1], [0, 1, 0, 2], [2, 0, 1, 2], [2, 0, 2], [2, 1, 2]];
  it('10 dollars when two identical non-consecutive symbols', () => {
    history.map(calcPrize).forEach(i => expect(i).toEqual(10));
  });

  history = [[1, 0, 1, 1], [0, 0, 2], [0, 1, 1], [2, 1, 2, 2]];
  it('20 dollars when two consecutive symbols', () => {
    history.map(calcPrize).forEach(i => expect(i).toEqual(20));
  });

  history = [[1, 1, 1], [0, 0, 0], [1, 1, 1, 1], [1, 1]];
  it('100 dollars when same symbol in all the wheels', () => {
    history.map(calcPrize).forEach(i => expect(i).toEqual(100));
  });
});
