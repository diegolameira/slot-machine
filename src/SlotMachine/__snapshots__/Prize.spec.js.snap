// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Prize 10 dollars when two identical non-consecutive symbols 1`] = `
Array [
  <h1>
    Prize: 
    0
  </h1>,
  <h2>
    Win Amount:
    0
  </h2>,
]
`;

exports[`Prize 20 dollars when two consecutive symbols 1`] = `
Array [
  <h1>
    Prize: 
    0
  </h1>,
  <h2>
    Win Amount:
    0
  </h2>,
]
`;

exports[`Prize 100 dollars when same symbol in all the wheels 1`] = `
Array [
  <h1>
    Prize: 
    0
  </h1>,
  <h2>
    Win Amount:
    0
  </h2>,
]
`;
